src/include/spinlock.h

```diff
@@ -18,52 +18,119 @@
 #define CEPH_SPINLOCK_HPP
 
 #include <atomic>
+#include <thread>
+
+#include <immintrin.h>
+
+#include "common/likely.h"
 
 namespace ceph {
 inline namespace version_1_0 {
 
 class spinlock;
 
-inline void spin_lock(std::atomic_flag& lock);
-inline void spin_unlock(std::atomic_flag& lock);
-inline void spin_lock(ceph::spinlock& lock);
-inline void spin_unlock(ceph::spinlock& lock);
+inline void spin_lock(std::atomic_bool& locked);
+inline void spin_unlock(std::atomic_bool& locked);
+inline void spin_lock(ceph::spinlock& locked);
+inline void spin_unlock(ceph::spinlock& locked);
+
+namespace spin {
+
+template <class MutexT, std::size_t MaxTriesV = 64>
+struct adapt_guard final
+{
+  typedef MutexT mutex_type;
+
+  adapt_guard(mutex_type& m)
+    : m(m)
+  {
+    if (likely(m.try_lock())) {
+      return;
+    }
+
+    std::size_t tries = MaxTriesV;
+    do {
+      _mm_pause();
+    } while (!m.try_lock() && --tries > 0);
+
+    if (!tries) {
+      m.lock();
+    }
+  }
+
+  adapt_guard(const adapt_guard&) = delete;
+
+  ~adapt_guard()
+  {
+    m.unlock();
+  }
+
+private:
+  mutex_type& m;
+};
+
+} // namespace spin
 
 /* A pre-packaged spinlock type modelling BasicLockable: */
 class spinlock final
 {
-  std::atomic_flag af = ATOMIC_FLAG_INIT;
+  // Not using atomic_flag anymore because it doesn't
+  // provide the load nor store operation.
+  std::atomic_bool locked = false;
+
+  // In contrast to atomic_flag, atomic_bool might be
+  // implemented on top of e.g. mutex. However, it is
+  // very unlikely we'll face such situation in Ceph.
+  static_assert(std::atomic_bool::is_always_lock_free);
 
-  public:
+public:
   void lock() {
-    ceph::spin_lock(af);
+    ceph::spin_lock(locked);
   }
  
   void unlock() noexcept {
-    ceph::spin_unlock(af);
+    ceph::spin_unlock(locked);
   }
 };
 
 // Free functions:
-inline void spin_lock(std::atomic_flag& lock)
+inline void spin_lock(std::atomic_bool& locked)
 {
- while(lock.test_and_set(std::memory_order_acquire))
-  ;
+  bool expected = false;
+  if (likely(locked.compare_exchange_weak(expected, true,
+                                          std::memory_order_acquire,
+                                          std::memory_order_relaxed))) {
+    return;
+  }
+
+  do {
+    std::size_t r3 = 0;
+    do {
+      asm("pause");
+      if (++r3 == 32) {
```

Comment:
TODO: make this compile-time configurable (including the `if  constexpr()` to disable yielding at all).

Reply:
FIXED.